sematic/api_client.py

```diff
@@ -97,15 +106,136 @@ def _put(endpoint, json_payload) -> Any:
     pass
 
 
+def _validate_server_compatibility(
+    tries: int = 5, seconds_between_tries: int = 10, use_cached: bool = True
+):
+    """Check that the client is compatible with the server.
+
+    Raises an error if the server and client are incompatible, or if this can't be
+    verified.
+    """
+    global _validated_client_version
+    if _validated_client_version and use_cached:
+        return
+
+    retriable_errors = (
+        ConnectionError,
+        ServerError,
+    )
+    response = None
+    error: Optional[Exception] = None
+    for _ in range(tries):
+        response, error = _validate_server_compatibility_no_catch()
+        if error is None:
+            _validated_client_version = True
+            return
+        if not isinstance(error, retriable_errors):
+            break
+        logger.warning(
+            "Unsuccessful at determining version compatibility waiting %s seconds...",
+            seconds_between_tries,
+        )
+        time.sleep(seconds_between_tries)
+    if response is not None and response.status_code != 200:
+        logger.error(
+            "Sematic server returned status code %s when"
+            " asked for version information. Response: %s",
+            response.status_code,
+            response.text,
+        )
+
+    assert error is not None  # should be impossible to fail, but makes mypy happy
+    raise error
+
+
+def _validate_server_compatibility_no_catch() -> Tuple[
+    Optional[requests.Response], Optional[Exception]
+]:
+    headers = {"Content-Type": "application/json"}
+    response = None
+    error: Optional[Exception] = None
+    try:
+        response = requests.get(_url("/meta/versions"), headers=headers)
+    except ConnectionError:
+        error = APIConnectionError(
+            (
+                f"Unable to connect to the Sematic API at "
+                f"{get_config().server_url}.\n Make sure the "
+                f"correct server address is set with\n"
+                f"\t$ sematic settings set "
+                f"{SettingsVar.SEMATIC_API_ADDRESS.value} <address>"
+            )
+        )
+        return None, error
+    unexpected_server_response_error = IncompatibleClientError(
+        "The Sematic server did not provide information about its version "
+        "in the expected format. It could be that the server is too old to "
+        "provide this information. Consider upgrading your server if possible, "
+        "or reach out to Sematic for support."
+    )
+    try:
+        response_json = response.json()
+    except JSONDecodeError:
+        return response, unexpected_server_response_error
+    if 400 <= response.status_code < 500:
+        return response, unexpected_server_response_error
+    if response.status_code >= 500:
+        error = ServerError(
+            f"The Sematic server is not responsive, and returned a status code of "
+            f"{response.status_code}."
+        )
+        return response, error
```

Comment:
All of this logic too needs to be centralized, or simply use `raise_for_status`. This is a lot of code for a simple version check 😬